### PR TITLE
refactoring to persistence.py #255

### PR DESCRIPTION
**Pull request by k9ert**
This is in preparation of #255. It's work in progress as the hook is not yet implemented.
Can wait for after the release.

### CODE REVIEW

**Comment by k9ert** [comment]
I guess this is now ready for review. I tried to get all cases where we're writing to ~/.specter. I searched for "open(", "shutil" and "os.remove".
I guess would fix now #255 .

File: src/cryptoadvance/specter/helpers.py [hunk 1]
```diff
@@ -69,83 +71,28 @@ def load_jsons(folder, key=None):
...
-                with open(os.path.join(folder, fname)) as f:
-                    d = json.load(f)
+            d = read_json_file(os.path.join(folder, fname))
             if key is None:
                 dd[fname[:-5]] = d
             else:
                 d["fullpath"] = os.path.join(folder, fname)
                 d["alias"] = fname[:-5]
                 dd[d[key]] = d
         except:
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/helpers.py:81]: `e` is not defined.
```suggestion
        except Exception as e:
```

File: src/cryptoadvance/specter/controller.py [hunk 2]
```diff
@@ -345,16 +347,7 @@ def general_settings():
...
-                with fslock:
-                    with open(
-                        os.path.join(
-                            app.specter.device_manager.data_folder,
-                            "%s.json" % device['alias']
-                        ),
-                        "w"
-                    ) as file:
-                        file.write(json.dumps(device, indent=4))
+            write_devices(devices)
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/controller.py:350]:
```suggestion
            write_devices(restore_devices)
```

Comment (k9ert) [reply src/cryptoadvance/specter/controller.py:350]: Great spot!

**Comment by k9ert** [comment]
Let me do revisit that the coming days before merging.

File: src/cryptoadvance/specter/persistence.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        with open(path, "w") as f:
+            json.dump(content, f, indent=4)
+
+
+def write_devices(devices):
+    """ write all the devices into the specter-folder """
+    with fslock:
+        with open(
+            os.path.join(
+                app.specter.device_manager.data_folder, "%s.json" % device["alias"]
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/persistence.py:50]: `app` is not defined
pass the path from where you use this function.

File: src/cryptoadvance/specter/persistence.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        ) as file:
+            file.write(json.dumps(device, indent=4))
+    storage_callback()
+
+
+def write_wallet(wallet_alias, wallet):
+    with fslock:
+        with open(
+            os.path.join(
+                app.specter.wallet_manager.working_folder, "%s.json" % wallet_alias
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/persistence.py:62]: Same, app is not defined.

**Comment by k9ert** [comment]
Hold on until i've created tests for restoring wallets as well.

**Comment by k9ert** [comment]
Ok, i guess this might be a lot closer to review :-).
I've tested the import and export. Seem to work.
I've created tests for this and now we can also see coverage of the tests. I was surprised to see that we have 42% test-coverage.

File: src/cryptoadvance/specter/controller.py [hunk 5]
```diff
@@ -385,15 +378,7 @@ def general_settings():
...
-                with fslock:
-                    with open(
-                        os.path.join(
-                            app.specter.wallet_manager.working_folder,
-                            "%s.json" % wallet['alias']
-                        ),
-                        "w"
-                    ) as file:
-                        file.write(json.dumps(wallet, indent=4))
+                write_wallet(wallet['alias'],wallet)
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/controller.py:381]: `TypeError: write_wallet() takes 1 positional argument but 2 were given`

**Comment by k9ert** [comment]
Hoping for the last round of review ...

**Pull request merged**